cmd/tether/attach_test.go

```diff
@@ -0,0 +1,613 @@
+
+	if !testServer.enabled {
+		t.Error("attach server was not enabled")
+		return
+	}
+
+	// create client on the mock pipe
+	conn, err := clientBackchannel(context.Background())
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	cconfig := &ssh.ClientConfig{
+		User: "daemon",
+		HostKeyCallback: func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+			return nil
+		},
+	}
+
+	// create the SSH client
+	sConn, chans, reqs, err := ssh.NewClientConn(conn, "notappliable", cconfig)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer sConn.Close()
+	client := ssh.NewClient(sConn, chans, reqs)
+
+	session, err := SSHAttach(client, testConfig.ID)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	stdout := session.Stdout()
+
+	// FIXME: this is line buffered - how do I disable that so we don't have odd hangs to diagnose
+	// when the trailing \n is missed
+	testBytes := []byte("hello world!\n")
+	// read from session into buffer
+	buf := &bytes.Buffer{}
+
+	var wg sync.WaitGroup
+	go func() { wg.Add(1); io.CopyN(buf, stdout, int64(len(testBytes))); wg.Done() }()
+
+	// write something to echo
+	log.Debug("sending test data")
+	session.Stdin().Write(testBytes)
+	log.Debug("sent test data")
+
+	// wait for the close to propogate
+	wg.Wait()
+	session.Stdin().Close()
+
+	if !bytes.Equal(buf.Bytes(), testBytes) {
+		t.Errorf("expected: \"%s\", actual: \"%s\"", string(testBytes), buf.String())
+		return
+	}
+}
+
+//
+/////////////////////////////////////////////////////////////////////////////////////
+
+/////////////////////////////////////////////////////////////////////////////////////
+// TestAttachTwoConfig sets up the config for attach testing - tests launching and
+// attaching to two different processes simultaneously
+//
+type TestAttachTwoConfig struct{}
+
+func (c *TestAttachTwoConfig) StoreConfig(*metadata.ExecutorConfig) (string, error) {
+	return "", errors.New("not implemented")
+}
+func (c *TestAttachTwoConfig) LoadConfig() (*metadata.ExecutorConfig, error) {
+	config := metadata.ExecutorConfig{}
+
+	config.ID = "attachtwo"
+	config.Name = "tether_test_executor"
+	config.Sessions = map[string]metadata.SessionConfig{
+		"tee1": metadata.SessionConfig{
+			Common: metadata.Common{
+				ID:   "tee1",
+				Name: "tether_test_session1",
+			},
+			Tty:    false,
+			Attach: true,
+			Cmd: metadata.Cmd{
+				Path: "/usr/bin/tee",
+				// grep, matching everything, reading from stdin
+				Args: []string{"/usr/bin/tee", pathPrefix + "/tee.out"},
+				Env:  []string{},
+				Dir:  "/",
+			},
+		},
+		"tee2": metadata.SessionConfig{
+			Common: metadata.Common{
+				ID:   "tee2",
+				Name: "tether_test_session2",
+			},
+			Tty:    false,
+			Attach: true,
+			Cmd: metadata.Cmd{
+				Path: "/usr/bin/tee",
+				// grep, matching everything, reading from stdin
+				Args: []string{"/usr/bin/tee", pathPrefix + "/tee2.out"},
+				Env:  []string{},
+				Dir:  "/",
+			},
+		},
+	}
+
+	// generate a host key for the tether
```

Comment:
It could be refactored, but I'm going to leave it as is for now for as it'll need changing in the near future once @caglar10ur has finished the first part of the configuration management and we start using that directly in the tether.

That may well result in having a set of base configurations that we can then modify and consume per test - that would make it easier to determine test intent from code, but harder to see total test config at a glance.